File: server/csharp/MetaObjects/AgentContext/ContentRoot.cs
// ContentRoot — resolve the agent-context/ content tree the assembler reads.
//
// Two resolution sources (matching the TS/Java/Python references' bundled-then-monorepo
// walk):
//   1. A bundled copy shipped alongside the assembly, at
//      <AppContext.BaseDirectory>/agent-context/ — the .NET tool / CLI csproj copies the
//      repo-root agent-context/ into its output via <Content Include CopyToOutputDirectory>.
//      This is the published path.
//   2. A dev fallback: walk up from AppContext.BaseDirectory (or an explicit anchor) to
//      the monorepo root and use its top-level agent-context/ directory.
//
// A directory is a valid content root iff it holds the authoring skill body.

namespace MetaObjects.AgentContext;

/// <summary>Resolves the <c>agent-context/</c> content tree the assembler reads.</summary>
public static class ContentRoot
{
    /// <summary>A directory is a valid content root iff it holds the authoring skill body.</summary>
    public static bool IsContentRoot(string directory) =>
        File.Exists(Path.Combine(directory, "skills", "metaobjects-authoring", "SKILL.md"));

    /// <summary>
    /// Resolve the content tree: prefer the bundled copy beside the assembly (published
    /// path); fall back to a monorepo <c>agent-context/</c> found by walking up from
    /// <paramref name="anchor"/> (or <see cref="AppContext.BaseDirectory"/> when null).
    /// </summary>
    /// <param name="anchor">A directory to start the monorepo walk-up from; defaults to the assembly base dir.</param>
    /// <returns>The resolved content root.</returns>
    /// <exception cref="DirectoryNotFoundException">If no valid content tree can be found.</exception>
    public static string Resolve(string? anchor = null)
    {
        var bundled = Path.Combine(AppContext.BaseDirectory, "agent-context");
        if (IsContentRoot(bundled)) return bundled;

        var dev = WalkUpForMonorepo(anchor ?? AppContext.BaseDirectory);
        if (dev is not null) return dev;

        throw new DirectoryNotFoundException(
            "agent-context content not found — looked for a bundled copy beside the assembly " +
            $"({bundled}) and a monorepo `agent-context/` walking up from " +
            (anchor ?? AppContext.BaseDirectory));
    }

    /// <summary>
    /// Walk up from <paramref name="start"/> to a directory holding a valid
    /// <c>agent-context/</c>, returning its path (or <c>null</c> if none is found).
    /// </summary>
    public static string? WalkUpForMonorepo(string start)
    {
        var dir = new DirectoryInfo(Path.GetFullPath(start));
        while (dir is not null)
        {
            var candidate = Path.Combine(dir.FullName, "agent-context");
            if (IsContentRoot(candidate)) return candidate;
            dir = dir.Parent;
        }
        return null;
    }
}

File: server/java/metadata/src/main/java/com/metaobjects/agentcontext/ContentRoot.java
package com.metaobjects.agentcontext;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.URISyntaxException;
import java.net.URL;
import java.nio.file.FileSystem;
import java.nio.file.FileSystemAlreadyExistsException;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Optional;

/**
 * Resolve the {@code agent-context/} content tree the assembler reads.
 *
 * <p>Two resolution sources (matching the TS/Python references' bundled-then-monorepo
 * walk):
 * <ol>
 *   <li>A bundled copy on the classpath at {@code /agent-context/} (the published
 *       plugin/jar carries this — see the module's {@code copy-resources} step that
 *       copies the repo-root {@code agent-context/} into {@code target/classes}).</li>
 *   <li>A dev fallback: walk up from a known anchor directory (e.g. the Maven
 *       project basedir or {@code user.dir}) to the monorepo root and use its
 *       top-level {@code agent-context/} directory.</li>
 * </ol>
 *
 * <p>A directory is a valid content root iff it holds the authoring skill body.
 */
public final class ContentRoot {

    private ContentRoot() {
    }

    /** A directory is a valid content root iff it holds the authoring skill body. */
    public static boolean isContentRoot(Path directory) {
        return Files.isRegularFile(
                directory.resolve("skills")
                        .resolve("metaobjects-authoring")
                        .resolve("SKILL.md"));
    }

    /**
     * Resolve the content tree, walking up from {@code anchor} for the dev monorepo
     * fallback, after first trying the classpath-bundled copy.
     *
     * @param anchor a directory to start the monorepo walk-up from (e.g. the Maven
     *               project basedir). May be {@code null} to use {@code user.dir}.
     * @return the resolved content root.
     * @throws IllegalStateException if no content tree can be found.
     */
    public static Path resolve(Path anchor) {
        Optional<Path> bundled = bundledOnClasspath();
        if (bundled.isPresent()) {
            return bundled.get();
        }
        Optional<Path> dev = walkUpForMonorepo(anchor);
        if (dev.isPresent()) {
            return dev.get();
        }
        throw new IllegalStateException(
                "agent-context content not found — looked for a classpath-bundled copy "
                        + "(/agent-context) and a monorepo `agent-context/` walking up from "
                        + (anchor != null ? anchor : System.getProperty("user.dir")));
    }

    /** The classpath-bundled content root ({@code /agent-context/}), if present + valid. */
    public static Optional<Path> bundledOnClasspath() {
        URL marker = ContentRoot.class.getClassLoader()
                .getResource("agent-context/skills/metaobjects-authoring/SKILL.md");
        if (marker == null) {
            return Optional.empty();
        }
        try {
            if ("file".equals(marker.getProtocol())) {
                Path skill = Paths.get(marker.toURI());
                // .../agent-context/skills/metaobjects-authoring/SKILL.md → .../agent-context
                Path root = skill.getParent().getParent().getParent();
                return isContentRoot(root) ? Optional.of(root) : Optional.empty();
            }
            if ("jar".equals(marker.getProtocol())) {
                Path root = jarContentRoot(marker);
                return isContentRoot(root) ? Optional.of(root) : Optional.empty();
            }
        } catch (URISyntaxException | IOException e) {
            throw new UncheckedIOException(
                    new IOException("agent-context: failed resolving bundled content from " + marker, e));
        }
        return Optional.empty();
    }

    /** Open (or reuse) the jar filesystem and return the {@code /agent-context} root inside it. */
    private static Path jarContentRoot(URL marker) throws IOException, URISyntaxException {
        // jar:file:/path/to.jar!/agent-context/skills/metaobjects-authoring/SKILL.md
        String spec = marker.toString();
        int bang = spec.indexOf("!/");
        String jarPart = spec.substring("jar:".length(), bang);
        Path jar = Paths.get(new URL(jarPart).toURI());
        FileSystem fs;
        try {
            fs = FileSystems.newFileSystem(jar, (ClassLoader) null);
        } catch (FileSystemAlreadyExistsException already) {
            fs = FileSystems.getFileSystem(java.net.URI.create("jar:" + jar.toUri()));
        }
        return fs.getPath("/agent-context");
    }

    /** Walk up from the anchor (or {@code user.dir}) to a dir holding a valid {@code agent-context/}. */
    public static Optional<Path> walkUpForMonorepo(Path anchor) {
        Path p = (anchor != null ? anchor : Paths.get(System.getProperty("user.dir")))
                .toAbsolutePath();
        while (p != null) {
            Path candidate = p.resolve("agent-context");
            if (isContentRoot(candidate)) {
                return Optional.of(candidate);
            }
            p = p.getParent();
        }
        return Optional.empty();
    }
}
